### PR TITLE
Show top artists and similar artists playlist (recommendations) on LB website

### PR DESCRIPTION
**Pull request by vansika**
# Problem
We have recently decided to store missing MusicBrainz data in the db and request users to submit the data so that the mapping and the recs can be improved, but until we show users what we have, until there is an incentive I think it is very unlikely for a user to submit data to MB. Also, user feedback would be a nice thing to have to improve recommendations.

# Solution
Show playlists (recommendations) on LB website 

# Action 
I have used some of the code from https://github.com/metabrainz/troi-recommendation-playground written by @mayhem to get the playlist on LB website

### CODE REVIEW

**Comment by vansika** [comment]
screenshots pending...

**Comment by vansika** [comment]
[Screenshot of user with top artist playlist ](https://drive.google.com/file/d/1bfmXOHEdCoZpGE-wtdo7dwRlf0BtB1es/view?usp=sharing)

[Screenshot of user without top artist playlist](https://drive.google.com/file/d/1yePpRMkPFq8Hdls4Lg7VABrkcFSbU2tO/view?usp=sharing)

**Comment by mayhem** [comment]
I'm glad you're taking the initiative to show this data on the site. However, there are two things that we should be careful with:

1. We should refrain from calling them playlists -- they aren't that. I would call them "recommendation candidate recordings" or something like that. If we call them playlists then people will be very unhappy with our work and that would end up being a poor first showing for us. There should be a clear statement that "these are tracks that being suggested as tracks that the user should also listen to, but that this isn't a finished product."
2. You're using bono from a production system -- we should only be using things from labs.api.listenbrainz.org . Though it won't be much work for me to move the end point and I'll happily do it, but we can't merge this until we're not pointing to bono anymore.

**Comment by vansika** [comment]
Can you move the endpoint so that I don't have to use bono anymore?

**Comment by mayhem** [comment]
Yes, will do. I can try to do that on the train tomorrow.

**Comment by vansika** [comment]
Oh. No hurry :)

**Comment by vansika** [comment]
There are still a few things left to be done on spark side then we can merge this pretty thing after all the requested and required changes are done :)

File: listenbrainz/webserver/templates/user/base.html
Comment (paramsingh) [inline listenbrainz/webserver/templates/user/base.html:13]: I don't think this needs to be on the user page right now. Let's make it a seperate URL, fix out usability and verbiage kinks over a few weeks and then see if we want to bring it front and center to a page that all users will see.

File: listenbrainz/webserver/views/test/test_recommendations_cf_recording.py
Comment (vansika) [inline listenbrainz/webserver/views/test/test_recommendations_cf_recording.py:142]: @paramsingh test fail here saying current_user doesn't exist. Can you have a look?

**Comment by vansika** [comment]
The return statement in [redis_connection.py](https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/webserver/redis_connection.py#L27) and [timescale_connection.py](https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/webserver/timescale_connection.py#L18) is not required since we are referring to these vars like this in the code 
`import webserver`
`webserver.redis_connection._redis`

scripts where we directly import `_redis` error out since we are actually importing them before initialization therefore 
`-redis=None`. The same thing is happening with `_ts`. 

either catch the return value and use it or reference the vars after initialization. Direct import is leading to errors.

File: listenbrainz/webserver/static/js/src/RecentListens.test.tsx
Comment (vansika) [inline listenbrainz/webserver/static/js/src/RecentListens.test.tsx:1189]: The commented tests fail for some reason which I don't understand.
Error: 

`Summary of all failing tests
 FAIL  static/js/src/RecentListens.test.tsx
  ● Test suite failed to run

    RangeError: Maximum call stack size exceeded

      at Math.<anonymous> (node_modules/jest-mock/build/index.js:379:51)
      at randomIntInRange (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:43:33)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:75:22)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
`

Comment (mayhem) [reply listenbrainz/webserver/static/js/src/RecentListens.test.tsx:1189]: This sounds like runaway recursion....

File: listenbrainz/webserver/static/js/src/RecentListens.test.tsx
Comment (vansika) [inline listenbrainz/webserver/static/js/src/RecentListens.test.tsx:1171]: I am only checking the length of returned array and not the contents since `array.slice` of the same array at different places is giving different results. 

I think the `listens` array is being modified somewhere in the code, or may be there is some other reason, I am not sure.

**Comment by vansika** [comment]
The PR is not as perfect as it should be. As I was writing it I realised that we should have a separate component for *recommendation* but writing a separate component is not in the scope of this PR I think. Maybe we can open a ticket for the new rec component and I will start working on it.

File: listenbrainz/webserver/views/user.py
Comment (mayhem) [inline listenbrainz/webserver/views/user.py]: Why is this in this PR?

**Review by mayhem — COMMENTED** [review]
The python bits all look sane, though I am not sure what delete listens is doing here? rebase issue? And while the tests for this feature are rather extensive, I wonder if writing extensive tests for a temporary feature is the best way to spend our time. (Also discussed in IRC)

File: listenbrainz/webserver/templates/recommendations_cf_recording/info.html
Comment (shivam-kapila) [inline listenbrainz/webserver/templates/recommendations_cf_recording/info.html:7]: this is loooooong

File: listenbrainz/webserver/templates/recommendations_cf_recording/info.html
Comment (shivam-kapila) [inline listenbrainz/webserver/templates/recommendations_cf_recording/info.html:19]: this too

**Review by shivam-kapila — CHANGES_REQUESTED** [review]

File: listenbrainz/webserver/static/js/src/RecentListens.tsx
Comment (ishaanshah) [inline listenbrainz/webserver/static/js/src/RecentListens.tsx:598]: Should this be `handleClickOldestRecommendations`?

File: listenbrainz/webserver/static/js/src/RecentListens.tsx
Comment (ishaanshah) [inline listenbrainz/webserver/static/js/src/RecentListens.tsx:668]: And this `handleClickNewestRecommendation`?

File: listenbrainz/webserver/static/js/src/RecentListens.tsx
Comment (ishaanshah) [inline listenbrainz/webserver/static/js/src/RecentListens.tsx:715]: Rather than creating 4 different handlers for each button we can pass the offset to a single function and remove lot of redundant code.

File: listenbrainz/webserver/static/js/src/RecentListens.tsx
Comment (ishaanshah) [inline listenbrainz/webserver/static/js/src/RecentListens.tsx:93]: `page` based pagination would be easier than `offset` I think. We use it on the Reports page, so it would be good for consistency too.

File: listenbrainz/webserver/static/js/src/RecentListens.tsx
Comment (vansika) [reply listenbrainz/webserver/static/js/src/RecentListens.tsx:94]: Ummm.... Like?

File: listenbrainz/webserver/static/js/src/RecentListens.test.tsx
Comment (vansika) [reply listenbrainz/webserver/static/js/src/RecentListens.test.tsx:1189]: Thanks :)

File: listenbrainz/webserver/static/js/src/RecentListens.tsx
Comment (vansika) [reply listenbrainz/webserver/static/js/src/RecentListens.tsx:639]: ` window.history.pushState(null, "", `?offset=${offset}`);` what is this doing exactly?

File: listenbrainz/webserver/templates/recommendations_cf_recording/info.html
Comment (vansika) [reply listenbrainz/webserver/templates/recommendations_cf_recording/info.html:7]: Long? This is the info page, it will be a lil long.

File: listenbrainz/webserver/static/js/src/RecentListens.tsx
Comment (vansika) [reply listenbrainz/webserver/static/js/src/RecentListens.tsx:94]: whoops. I read it mock. lol

**Review by paramsingh — APPROVED** [review]
Happy to merge this given that RecentListens needs to be refactored anyways

**Review by shivam-kapila — APPROVED** [review]
Happy to merge this. We will need to reflect page no. in url to jump directly to that page. Can be addressed in future PR.

**Review by ishaanshah — APPROVED** [review]
LGTM :)

**Comment by vansika** [comment]
@paramsingh @mayhem I think this should be merged on Monday after the first weekly recommendation job.

File: listenbrainz/webserver/views/recommendations_cf_recording.py
Comment (mayhem) [inline listenbrainz/webserver/views/recommendations_cf_recording.py:16]: Ideally this should be a better name -- server for what??

**Review by mayhem — APPROVED** [review]
I think the server name var needs a better name -- but take care of that after the release today. Lets keep moving!

File: listenbrainz/webserver/views/recommendations_cf_recording.py
Comment (vansika) [reply listenbrainz/webserver/views/recommendations_cf_recording.py:16]: Cool

**Pull request merged**
